resources/constraint_templates/lock_priv_capabilities.yaml

```diff
@@ -0,0 +1,120 @@
+apiVersion: templates.gatekeeper.sh/v1
+kind: ConstraintTemplate
+metadata:
+  name: lockprivcapabilities
+  annotations:
+    metadata.gatekeeper.sh/title: "Lock capabilities for namespaces with psa - PSA"
+    metadata.gatekeeper.sh/version: 1.0.0
+    description: >-
+      Controls Linux capabilities on containers on psa labels with privileged. This is to ensure the namespaces are locked down to the same level as they were when custom PSPs were implemented. This corresponds to the
+      `allowedCapabilities` and `requiredDropCapabilities` fields in a PodSecurityPolicy.
+spec:
+  crd:
+    spec:
+      names:
+        kind: LockPrivCapabilities
+      validation:
+        openAPIV3Schema:
+          type: object
+          description: >-
+            Controls Linux capabilities on containers. Corresponds to the
+            `allowedCapabilities` and `requiredDropCapabilities` fields in a
+            PodSecurityPolicy. For more information, see
+            https://kubernetes.io/docs/concepts/policy/pod-security-policy/#capabilities
+          properties:
+            allowedCapabilities:
+              type: array
+              description: A list of Linux capabilities that can be added to a container.
+              items:
+                type: string
+            requiredDropCapabilities:
+              type: array
+              description: A list of Linux capabilities that are required to be dropped from a container.
+              items:
+                type: string
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package lockprivcapabilities
+
+        import data.lib.exclude_update.is_update
+
+        violation[{"msg": msg}] {
+          # spec.containers.securityContext.capabilities field is immutable.
+          not is_update(input.review)
+
+          container := input.review.object.spec.containers[_]
+          has_disallowed_capabilities(container)
+          msg := sprintf("container <%v> has a disallowed capability. Allowed capabilities are %v", [container.name, get_default(input.parameters, "allowedCapabilities", "NONE")])
+        }
+
+        violation[{"msg": msg}] {
+          not is_update(input.review)
```

Comment:
Why are we violating when it is not an update? Does this allow a user to escalate privileges by creating one set of capabilities and then updating to another?

Reply:
My understanding is, it only violate when a new pod is created with a spec which is missing/allowed as specified. This will allow any existing pods running with these capabilities missing/allowed and hence allow migration. Also, this field is immutable and therefore cant be changed by updating the pod and refered in here: https://github.com/open-policy-agent/gatekeeper-library/commit/bd942ad8afd316743fbe97a69c5019d349c1a5c1